sql: failure running SQL benchmarks

```
panic: zero transaction timestamp in EvalContext

goroutine 13571 [running]:
panic(0x4edb180, 0xc856f44bb0)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/cockroachdb/cockroach/sql/parser.(*EvalContext).GetTxnTimestamp(0xc849f62fc0, 0x0, 0x0)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:777 +0xa8
github.com/cockroachdb/cockroach/sql/parser.glob.func91(0x1, 0xece87ee50, 0x234836b7, 0x5df39a0, 0x0, 0x0, 0xc842e15ec0, 0xc828526350, 0x0, 0x0, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/parser/builtins.go:1098 +0x54
github.com/cockroachdb/cockroach/sql/parser.(*FuncExpr).Eval(0xc82a952480, 0x1, 0xece87ee50, 0x234836b7, 0x5df39a0, 0x0, 0x0, 0xc842e15ec0, 0xc828526350, 0x0, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:1185 +0x53f
github.com/cockroachdb/cockroach/sql/parser.evalExprs(0x1, 0xece87ee50, 0x234836b7, 0x5df39a0, 0x0, 0x0, 0xc842e15ec0, 0xc828526350, 0x0, 0x0, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:1428 +0x138
github.com/cockroachdb/cockroach/sql/parser.(*Tuple).Eval(0xc849f63550, 0x1, 0xece87ee50, 0x234836b7, 0x5df39a0, 0x0, 0x0, 0xc842e15ec0, 0xc828526350, 0x0, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/parser/eval.go:1444 +0x7b
github.com/cockroachdb/cockroach/sql.(*planner).ValuesClause(0xc835346078, 0xc84123dae0, 0x0, 0x0, 0x1)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/values.go:70 +0x8b3
github.com/cockroachdb/cockroach/sql.(*planner).makePlan(0xc835346078, 0x7219480, 0xc84123dae0, 0x7219400, 0x0, 0x0, 0xa)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/plan.go:268 +0xb99
github.com/cockroachdb/cockroach/sql.(*planner).Insert(0xc835346078, 0xc856f78780, 0xe984f100, 0x0, 0x0, 0x51a6ea8)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/insert.go:116 +0x9cd
github.com/cockroachdb/cockroach/sql.(*planner).makePlan(0xc835346078, 0x7219338, 0xc856f78780, 0xc82644cc00, 0x0, 0x0, 0x12e4)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/plan.go:220 +0x1313
github.com/cockroachdb/cockroach/sql.(*Executor).execStmt(0xc8213562a0, 0x7219338, 0xc856f78780, 0xc835346078, 0xece87ee50, 0xc8234843c6, 0x5df39a0, 0x0, 0x0, 0x0, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/executor.go:867 +0x89
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtInOpenTxn(0xc8213562a0, 0x7219338, 0xc856f78780, 0xc835346078, 0x0, 0x143f44946e8d56b7, 0x0, 0xc835346028, 0x0, 0x0, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/executor.go:774 +0x542
github.com/cockroachdb/cockroach/sql.(*Executor).execStmtsInCurrentTxn(0xc8213562a0, 0xc821f64d00, 0x7, 0x8, 0xc835346078, 0xc835346028, 0x100, 0x0, 0x0, 0x0, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/executor.go:556 +0x630
github.com/cockroachdb/cockroach/sql.runTxnAttempt(0xc8213562a0, 0xc835346078, 0x1, 0xc835346028, 0xc84123dfa0, 0xc821f64d00, 0x7, 0x8, 0xc849f65448, 0xc849f65460, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/executor.go:488 +0x147
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest.func1(0xc82814b2c0, 0xc84123dfa0, 0xc82814b2c0)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/executor.go:401 +0x325
github.com/cockroachdb/cockroach/client.(*Txn).Exec(0xc82814b2c0, 0x52a0001, 0x143f44946b262293, 0x0, 0xc849f65558, 0x0)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/client/txn.go:495 +0x410
github.com/cockroachdb/cockroach/sql.(*Executor).execRequest(0xc8213562a0, 0xc835346000, 0xc856f77000, 0x17e, 0x0, 0x0, 0x0, 0x1000)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/executor.go:405 +0x8a5
github.com/cockroachdb/cockroach/sql.(*Executor).ExecuteStatements(0xc8213562a0, 0xc835346000, 0xc856f77000, 0x17e, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/executor.go:294 +0xbe
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).executeStatements(0xc82096a900, 0xc856f77000, 0x17e, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x2a952201, ...)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:592 +0x98
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).handleSimpleQuery(0xc82096a900, 0xc82096a928, 0x0, 0x0)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:303 +0xcd
github.com/cockroachdb/cockroach/sql/pgwire.(*v3Conn).serve(0xc82096a900, 0xc8485486e0, 0x0, 0x0)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/pgwire/v3.go:263 +0xf8a
github.com/cockroachdb/cockroach/sql/pgwire.(*Server).ServeConn(0xc82047a058, 0x7219e70, 0xc848888600, 0x0, 0x0)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/sql/pgwire/server.go:157 +0x8df
github.com/cockroachdb/cockroach/server.(*Server).Start.func4.1(0x7219918, 0xc826392160)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/server/server.go:300 +0x4b
github.com/cockroachdb/cockroach/util.ServeHandler.func3.1(0xc824b51700, 0x7219918, 0xc826392160, 0xc84c2b6010)
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/util/net.go:125 +0x5a
created by github.com/cockroachdb/cockroach/util.ServeHandler.func3
	/Users/spencerkimball/go/src/github.com/cockroachdb/cockroach/util/net.go:127 +0x31f
```

--- a/client/txn.go
+++ b/client/txn.go
@@ -514,20 +514,20 @@ func (txn *Txn) Exec(
 		panic("asked to retry or commit a txn that is already aborted")
 	}
 
-	if txn != nil {
-		// If we're looking at a brand new transaction, then communicate
-		// what should be used as initial timestamp for the KV txn created
-		// by TxnCoordSender.
-		if txn.Proto.OrigTimestamp == roachpb.ZeroTimestamp {
-			txn.Proto.OrigTimestamp = opt.MinInitialTimestamp
-		}
-	}
-
 	if opt.AutoRetry {
 		retryOptions = txn.db.txnRetryOptions
 	}
 RetryLoop:
 	for r := retry.Start(retryOptions); r.Next(); {
+		if txn != nil {
+			// If we're looking at a brand new transaction, then communicate
+			// what should be used as initial timestamp for the KV txn created
+			// by TxnCoordSender.
+			if txn.Proto.OrigTimestamp == roachpb.ZeroTimestamp {
+				txn.Proto.OrigTimestamp = opt.MinInitialTimestamp
+			}
+		}
+
 		pErr = fn(txn, &opt)
 		if txn != nil {
 			txn.retrying = true

--- a/client/txn_test.go
+++ b/client/txn_test.go
@@ -490,6 +490,46 @@ func TestRunTransactionRetryOnErrors(t *testing.T) {
 	}
 }
 
+// TestAbortedRetryPreservesTimestamp verifies that the minimum timestamp
+// set by the client is preserved across retries of aborted transactions.
+func TestAbortedRetryPreservesTimestamp(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	// Create a TestSender that aborts a transaction 2 times before succeeding.
+	count := 0
+	db := newDB(newTestSender(func(ba roachpb.BatchRequest) (*roachpb.BatchResponse, *roachpb.Error) {
+		if _, ok := ba.GetArg(roachpb.Put); ok {
+			count++
+			if count < 3 {
+				return nil, roachpb.NewError(&roachpb.TransactionAbortedError{})
+			}
+		}
+		return ba.CreateReply(), nil
+	}, nil))
+
+	txnClosure := func(txn *Txn, opt *TxnExecOptions) *roachpb.Error {
+		// Ensure the KV transaction is created.
+		return txn.Put("a", "b")
+	}
+
+	txn := NewTxn(context.Background(), *db)
+
+	// Request a client-defined timestamp.
+	refTimestamp := roachpb.Timestamp{WallTime: 42, Logical: 69}
+	execOpt := TxnExecOptions{AutoRetry: true, AutoCommit: true}
+	execOpt.MinInitialTimestamp = refTimestamp
+
+	// Perform the transaction.
+	if pErr := txn.Exec(execOpt, txnClosure); pErr != nil {
+		t.Fatal(pErr)
+	}
+
+	// Check the timestamp was preserved.
+	if txn.Proto.OrigTimestamp != refTimestamp {
+		t.Errorf("expected txn orig ts to be %s; got %s", refTimestamp, txn.Proto.OrigTimestamp)
+	}
+}
+
 // TestAbortTransactionOnCommitErrors verifies that transactions are
 // aborted on the correct errors.
 func TestAbortTransactionOnCommitErrors(t *testing.T) {

--- a/sql/executor.go
+++ b/sql/executor.go
@@ -40,7 +40,6 @@ import (
 	"github.com/cockroachdb/cockroach/util/metric"
 	"github.com/cockroachdb/cockroach/util/retry"
 	"github.com/cockroachdb/cockroach/util/stop"
-	"github.com/cockroachdb/cockroach/util/timeutil"
 )
 
 var errNoTransactionInProgress = errors.New("there is no transaction in progress")
@@ -774,8 +773,7 @@ func (e *Executor) execStmtInOpenTxn(
 	if txnState.tr != nil {
 		txnState.tr.LazyLog(stmt, true /* sensitive */)
 	}
-	result, pErr := e.execStmt(stmt, planMaker, timeutil.Now(),
-		implicitTxn /* autoCommit */)
+	result, pErr := e.execStmt(stmt, planMaker, implicitTxn /* autoCommit */)
 	if pErr != nil {
 		if txnState.tr != nil {
 			txnState.tr.LazyPrintf("ERROR: %v", pErr)
@@ -870,7 +868,7 @@ func commitSQLTransaction(txnState *txnState, p *planner,
 // the current transaction might have been committed/rolled back when this returns.
 func (e *Executor) execStmt(
 	stmt parser.Statement, planMaker *planner,
-	timestamp time.Time, autoCommit bool) (Result, *roachpb.Error) {
+	autoCommit bool) (Result, *roachpb.Error) {
 	var result Result
 	plan, pErr := planMaker.makePlan(stmt, autoCommit)
 	if pErr != nil {

--- a/sql/txn_restart_test.go
+++ b/sql/txn_restart_test.go
@@ -215,7 +215,7 @@ func TestTxnRestart(t *testing.T) {
 
 	if _, err := sqlDB.Exec(`
 CREATE DATABASE t;
-CREATE TABLE t.test (k TEXT PRIMARY KEY, v TEXT);
+CREATE TABLE t.test (k TEXT PRIMARY KEY, v TEXT, t DECIMAL);
 `); err != nil {
 		t.Fatal(err)
 	}
@@ -229,6 +229,9 @@ CREATE TABLE t.test (k TEXT PRIMARY KEY, v TEXT);
 		"hooly":     2,
 		"josephine": 2,
 		"laureal":   2,
+	}
+	magicVals.abortCounts = map[string]int{
+		"boulanger": 2,
 	}
 	magicVals.endTxnRestartCounts = map[string]int{
 		"boulanger": 2,
@@ -246,16 +249,18 @@ CREATE TABLE t.test (k TEXT PRIMARY KEY, v TEXT);
 
 	// Test that implicit txns - txns for which we see all the statements and prefixes
 	// of txns (statements batched together with the BEGIN stmt) - are retried.
+	// We also exercise the SQL cluster logical timestamp in here, because
+	// this must be properly propagated across retries.
 	if _, err := sqlDB.Exec(`
-INSERT INTO t.test (k, v) VALUES ('a', 'boulanger');
+INSERT INTO t.test (k, v, t) VALUES ('a', 'boulanger', cluster_logical_timestamp());
 BEGIN;
-INSERT INTO t.test (k, v) VALUES ('c', 'dromedary');
-INSERT INTO t.test (k, v) VALUES ('e', 'fajita');
+INSERT INTO t.test (k, v, t) VALUES ('c', 'dromedary', cluster_logical_timestamp());
+INSERT INTO t.test (k, v, t) VALUES ('e', 'fajita', cluster_logical_timestamp());
 END;
-INSERT INTO t.test (k, v) VALUES ('g', 'hooly');
+INSERT INTO t.test (k, v, t) VALUES ('g', 'hooly', cluster_logical_timestamp());
 BEGIN;
-INSERT INTO t.test (k, v) VALUES ('i', 'josephine');
-INSERT INTO t.test (k, v) VALUES ('k', 'laureal');
+INSERT INTO t.test (k, v, t) VALUES ('i', 'josephine', cluster_logical_timestamp());
+INSERT INTO t.test (k, v, t) VALUES ('k', 'laureal', cluster_logical_timestamp());
 `); err != nil {
 		t.Fatal(err)
 	}
@@ -300,7 +305,7 @@ BEGIN;
 	}
 
 	// Continue the txn in a new request, which is not retriable.
-	_, err := sqlDB.Exec("INSERT INTO t.test (k, v) VALUES ('g', 'hooly')")
+	_, err := sqlDB.Exec("INSERT INTO t.test (k, v, t) VALUES ('g', 'hooly', cluster_logical_timestamp())")
 	if !testutils.IsError(
 		err, "encountered previous write with future timestamp") {
 		t.Errorf("didn't get expected injected error. Got: %s", err)

